firebase-crashlytics/src/main/java/com/google/firebase/crashlytics/core/CrashlyticsController.java

```diff
@@ -1809,83 +1805,47 @@ private void sendSessionReports(AppSettingsData appSettings, boolean dataCollect
-        eventLatch.countDown();
-      }
-    }
-  }
-
   /**
    * Send an App Exception event to Firebase Analytics. FA records the event asynchronously, so this
    * method returns a Task in case the caller wants to verify that the event was recorded by FA and
    * will not be lost.
    */
   private Task<Void> recordFatalFirebaseEvent(long timestamp) {
-    final ThreadPoolExecutor executor = new ScheduledThreadPoolExecutor(1);
-    return Tasks.call(
-        executor,
-        new Callable<Void>() {
-          @Override
-          public Void call() throws Exception {
-            if (firebaseCrashExists()) {
-              Logger.getLogger()
-                  .d(
-                      Logger.TAG,
-                      "Skipping logging Crashlytics event to Firebase, FirebaseCrash exists");
-              return null;
-            }
-            if (analyticsConnector == null) {
-              Logger.getLogger()
-                  .d(
-                      Logger.TAG,
-                      "Skipping logging Crashlytics event to Firebase, no Firebase Analytics");
-              return null;
-            }
-            final BlockingCrashEventListener blockingListener = new BlockingCrashEventListener();
-            analyticsReceiver.setCrashlyticsOriginEventListener(blockingListener);
-
-            Logger.getLogger().d(Logger.TAG, "Logging Crashlytics event to Firebase");
-            final Bundle params = new Bundle();
-            params.putInt(FIREBASE_CRASH_TYPE, FIREBASE_CRASH_TYPE_FATAL);
-            params.putLong(FIREBASE_TIMESTAMP, timestamp);
 
-            analyticsConnector.logEvent(
-                FIREBASE_ANALYTICS_ORIGIN_CRASHLYTICS, FIREBASE_APPLICATION_EXCEPTION, params);
+    if (firebaseCrashExists()) {
+      Logger.getLogger()
+          .d(Logger.TAG, "Skipping logging Crashlytics event to Firebase, FirebaseCrash exists");
+      return Tasks.forResult(null);
+    }
+    if (analyticsConnector == null) {
+      Logger.getLogger()
+          .d(Logger.TAG, "Skipping logging Crashlytics event to Firebase, no Firebase Analytics");
+      return Tasks.forResult(null);
+    }
 
-            blockingListener.awaitEvent();
+    final TaskCompletionSource<Void> source = new TaskCompletionSource<>();
+    // This listener completes the TaskCompletionSource when it receives any app exception event
+    // that originated from Crashlytics. The implementation assumes there can be at most ONE app
+    // exception event being processed and waited upon at a time.
+    analyticsReceiver.setCrashlyticsOriginEventListener(
+        (id, extras) -> {
+          String eventName = extras.getString(AnalyticsConnectorReceiver.EVENT_NAME_KEY);
+          if (AnalyticsConnectorReceiver.APP_EXCEPTION_EVENT_NAME.equals(eventName)) {
+            Logger.getLogger().d(Logger.TAG, "Received app exception event callback from FA");
+            source.setResult(null);
```

Comment:
One concern I have here: This blocks indefinitely, so if for some reason we never get the analytics callback, the app will hang rather than terminate. Ideas?

Reply:
This should ultimately be handled by the timeout for the uncaught exception handler chain, no?

Reply:
Also, you should use `trySetResult` to avoid the possibility of throwing an exception here.

Reply:
The callback issue turned out to be due to a pre-existing issue in the awaitEvenIfOnMainThread impl, which is fixed in the newest commit on this PR.